### PR TITLE
Added Tx20 Wind Sensor

### PR DESCRIPTION
**Pull request by TheEggi**
Details of the sensor can be found in this issue:
https://github.com/arendst/Sonoff-Tasmota/issues/2654

The wind sensor consist of:
- red cable (vcc)
- yellow (gnd)
- black cable (transmit) - can be configured using UI
- green cable (starts transmission - should be connected to gnd)

Config:
image

Result:
image

Json Result:
"TX20":{"SPEED_AVG:":nan, "SPEED_MAX":0.00, "DIRECTION":"NNE"}

### CODE REVIEW

**Comment by TheEggi** [comment]
To be honest ... I don't have anything that I could compare it to. I mean the position is marked on the sensor and seems to work as expected but for the wind speed the only test that I could do was to blow at it and see if the speed is increasing/decreasing - and that seemed to have worked. If you have a better way/idea (without any real testing equipment) how I can test the precision of the device, I am more than happy to try it.

**Comment by Jason2866** [comment]
I think i buy one. I can compare it with the quality (expensive) Aneometer from my brother (hang gliding...)
Thx

**Comment by Silencer80** [comment]
Thank you very much @TheEggi for implementing this! 

I have just returned from a businesstrip and have a lot of work to do. However, I already have a NodeMCU and a TX20 sensor on my desk to test that as soon as possible. 

I'm so curious if that works well. I can compare the results with my mdt SCN-WS3HW.01 Weatherstation.

Thank you again!
-Boris

**Comment by FabianHM** [comment]
Hello, 
What are the chances that Tasmota's support will be short-term? I could buy the device just under 30 € but would not buy it without Tasmota use.
Greetings Fabian

**Comment by Jason2866** [comment]
@FabianHM 
Dont hesitate. Buy one it works with the fork from @TheEggi

File: sonoff/sonoff.ino [hunk 1]
```diff
@@ -2551,4 +2551,5 @@ void loop()
 
 //  yield();     // yield == delay(0), delay contains yield, auto yield in loop
   delay(sleep);  // https://github.com/esp8266/Arduino/issues/2021
+  XsnsCall(FUNC_LOOP);
```
Comment (ascillato) [inline sonoff/sonoff.ino]: Hi, This line shouldn't be there.

Please, check other sensors to see the way tasmota handle this.

File: sonoff/xsns_91_tx20.ino [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
+
+/*********************************************************************************************\
+ * Tx20 wind sensor
+ \*********************************************************************************************/
+
```
Comment (ascillato) [inline sonoff/xsns_91_tx20.ino]: You need to add a 
```suggestion
#ifdef USE_TX20_wind_sensor
```
in order to enable or disable this sensor from user_config.h file

By default this sensor should be disabled in user_config.h

File: sonoff/xsns_91_tx20.ino [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+      Tx20Show(1);
+      break;
+#ifdef USE_WEBSERVER
+    case FUNC_WEB_APPEND:
+      Tx20Show(0);
+      break;
+#endif  // USE_WEBSERVER
+  }
+  return result;
+}
```
Comment (ascillato) [inline sonoff/xsns_91_tx20.ino]: Add below this a
#endif // enable sensor

**Review by ascillato — CHANGES_REQUESTED** [review]

**Comment by ascillato** [comment]
@TheEggi 

This PR has been rebased with the new version of Tasmota.

Could you update it?

**Comment by FabianHM** [comment]
> 
> 
> Details of the sensor can be found in this issue:
> #2654
> 
> The wind sensor consist of:
> 
>     * red cable (vcc)
> 
>     * yellow (gnd)
> 
>     * black cable (transmit) - can be configured using UI
> 
>     * green cable (starts transmission - should be connected to gnd)
> 
> 
> Config:
> image
> 
> Result:
> image
> 
> Json Result:
> "TX20":{"SPEED_AVG:":nan, "SPEED_MAX":0.00, "DIRECTION":"NNE"}

What does the analogue input in screenshot "10" indicate?

**Comment by ascillato** [comment]
Means that in his version he has enabled the use of the Analog input on the pin A0 but nothing is connected there. Just that. Nothing to do with this PR nor the sensor. Don't worry.

**Comment by TheEggi** [comment]
Added most of your suggestions. Maybe you have a suggestion on which sensor I could check as an example to remove the FUNC_LOOP.

**Comment by Jason2866** [comment]
Hello together,
i have bought a TX20 this week. Happy to test when the "final" version is done.

**Comment by arendst** [comment]
Just released a version for the TX20. Give it a try and let me know if it works pls.

**Comment by TheEggi** [comment]
Seems to be working. Thank you for adding it.

**Comment by Jason2866** [comment]
Thx for programming and adding.
image

**Comment by arendst** [comment]
Thx for testing ;-)

**Comment by Sepultura22** [comment]
Hi, I have a doubt. I bought a TX20 sensor and it does not work for me. I have loaded the latest tastmota firmware in a basic sonoff and soldered the black cable in gpio4, the yellow and green cable in gnd and the red in vcc. I have also selected in the configuration that gpio4 acts with TX20. Should we do more? Thank you very much for your work, it is just what I was looking for.

**Comment by travelling-man** [comment]
Hello,

I have a problem that only updates the wind direction. The speed doesn't change and stays at 0.0 km/h permanently.

Does anyone have similar problems?

**Comment by Jason2866** [comment]
Nope, works fine for me

**Comment by FabianHM** [comment]
it works fine for me too. No Problems

**Comment by travelling-man** [comment]
what hardware do you use?

All cables (red -> 3,3v) (green, red -> gnd) (black -> d2) are already connected before starting the Wemos.

I am using a Wemos d1 mini and its 3,3v.

Strange is that the wind direction is updated.

Is there a way to debug it?

**Comment by Jason2866** [comment]
Just this...

**Comment by Jason2866** [comment]
Isnt installed because solar charging part isnt done right now...

**Comment by travelling-man** [comment]
> I cut the cable down to ~1.5m as I did not need it to be any longer - worked without any issues.

Can you take a picture with the model number?

**Comment by travelling-man** [comment]
Does anyone still have a photo with the original pin assignment?

**Comment by davide5681** [comment]
e possibile direttente da taota abilitare un us ita una volta arrivati ad una sohlia di vrlocita?

**Comment by davide5681** [comment]
e possibile direttamente da tasmota abilitare un uscita una volta arrivati ad una soglia  di velocità ?

**Comment by Pastant** [comment]
I tried the TX23 because the TX20 is not available any longer. The TX23 is not compatible! Even the support for the TX20 was dropped in recent releases (8.1) ...

**Comment by Jason2866** [comment]
Support for TX20 is NOT dropped. It is just not in a precompiled version (as many other drivers)
Read section self compile in the documentation www.tasmota.com

**Comment by Pastant** [comment]
The "Complete list of available feature and sensors." states that tasmota-sensors.bin supports the TX20. So the precompiled supprt is dropped. I won't compile a version because I only own a TX23 which is not compatible ...

**Comment by curzon01** [comment]
I'm currently trying to adapt the TX20 driver also for TX23. The protocol is the same but TX23 needs an initial signal to send out their data where TX20 sends every 2 seconds automatically.

**Pull request closed without merging**